### PR TITLE
282 discover all nested objects

### PR DESCRIPTION
**Pull request by olegbaturin**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌

#282 
#286 

1. Add tests for nested structures.
2. Update `buildObjectsCache()` to discover all nested objects only if required.
4.1 Update `dumpNestedInternal()` to stop traverse deeper if object is not found in the cache. This prevents possible errors in the future.
4.2 Fix `dumpNestedInternal()` to not stop dumping objects on the level 2*$depth.

### CODE REVIEW

**Review by vjik — COMMENTED** [review]
Yet another test:

```php
public function testAsJsonObjectsMapWithNestedObject(): void
{
	$nested = new stdClass();
	$nested->name = 'nested';
	$nestedId = spl_object_id($nested);

	$object = new stdClass();
	$object->name = 'root';
	$object->var = $nested;
	$objectId = spl_object_id($object);

	$this->assertSame(
		<<<JSON
		{
			"stdClass#$objectId": {
				"public \$name": "root",
				"public \$var": "stdClass#$nestedId (...)"
			}
		}
		JSON,
		Dumper::create($object)->asJsonObjectsMap(1, true)
	);
}
```

**Comment by olegbaturin** [comment]
> Yet another test:

Both objects must be present in the expected result.

**Comment by vjik** [comment]
> Both objects must be present in the expected result.

No. Depth is 1, so second object don't must be in result.

**Comment by olegbaturin** [comment]
> > Both objects must be present in the expected result.
> 
> No. Depth is 1, so second object don't must be in result.

It's a trap.

https://github.com/yiisoft/yii-debug/issues/282#issuecomment-2454144914

File: src/Dumper.php [hunk 1]
```diff
@@ -47,42 +51,44 @@ public function asJson(int $depth = 50, bool $format = false): string
 
     /**
      * Export variable as JSON summary of topmost items.
+     * Dumper go into the variable on full depth for search all objects.
```
Comment (samdark) [inline src/Dumper.php:54]:
```suggestion
     * Dumper goes into the variable full depth to search all objects.
```

**Comment by olegbaturin** [comment]
Explanaition of `buildObjectsCache()`

In this PR, the `buildObjectsCache()` creates different `$this->objects` caches for `asJsonObjectsMap()` and `asJson()` methods.
For the `asJsonObjectsMap()` method with @vjik's improvements, it caches all possible objects from the $variable.

So this condition in the `dumpNestedInternal()` will never fire and can be removed indeed.
https://github.com/olegbaturin/yii3-debug/blob/7ca18453c342f027fa0c4bd7ab55c42b873ce3f0/src/Dumper.php#L174

Additional calls of the `buildObjectsCache()` from the `dumpNestedInternal()` also have no sense.

**Comment by olegbaturin** [comment]
Proposal for `Callables` dumping

`Callables` are dumped as strings and not as objects.
https://github.com/olegbaturin/yii3-debug/blob/7ca18453c342f027fa0c4bd7ab55c42b873ce3f0/src/Dumper.php#L159

But in the `$this->objects` cache `Callable` objects are still presented, and they are dumped without links to them from the parent properties.

```
"Yiisoft\\Definitions\\CallableDefinition#149": {
	"private $callable": "static fn (\\Yiisoft\\Aliases\\Aliases $aliases) => new \\Yiisoft\\Mailer\\MessageBodyTemplate(\n    $aliases->get($params['yiisoft\/mailer']['messageBodyTemplate']['viewPath']),\n)"
},
"Closure#147": "static fn (\\Yiisoft\\Aliases\\Aliases $aliases) => new \\Yiisoft\\Mailer\\MessageBodyTemplate(\n    $aliases->get($params['yiisoft\/mailer']['messageBodyTemplate']['viewPath']),\n)",
```

The proposal is to skip `Callable` objects in the `buildObjectsCache()` method.

```
"Yiisoft\\Definitions\\CallableDefinition#149": {
	"private $callable": "static fn (\\Yiisoft\\Aliases\\Aliases $aliases) => new \\Yiisoft\\Mailer\\MessageBodyTemplate(\n    $aliases->get($params['yiisoft\/mailer']['messageBodyTemplate']['viewPath']),\n)"
},
```

**Comment by vjik** [comment]
> Proposal for `Callables` dumping

Can we make it in separate PR?

**Comment by samdark** [comment]
Is it good to go?

**Comment by vjik** [comment]
After discussion with @olegbaturin, we have suggestion.

1) Rename `$depth` parameter in `asJsonObjectsMap()` method to `$arrayDepth`.
2) `$arrayDepth` will be mean output depth for array object properties only.
3) And therefore, pass `$depth+2` instead of `$depth+1` in `asJsonInternal()`.

@xepozz WDYT?

**Comment by xepozz** [comment]
> Proposal for `Callables` dumping
> 
> `Callables` are dumped as strings and not as objects. https://github.com/olegbaturin/yii3-debug/blob/7ca18453c342f027fa0c4bd7ab55c42b873ce3f0/src/Dumper.php#L159
> 
> But in the `$this->objects` cache `Callable` objects are still presented, and they are dumped without links to them from the parent properties.
> 
> ```
> "Yiisoft\\Definitions\\CallableDefinition#149": {
> 	"private $callable": "static fn (\\Yiisoft\\Aliases\\Aliases $aliases) => new \\Yiisoft\\Mailer\\MessageBodyTemplate(\n    $aliases->get($params['yiisoft\/mailer']['messageBodyTemplate']['viewPath']),\n)"
> },
> "Closure#147": "static fn (\\Yiisoft\\Aliases\\Aliases $aliases) => new \\Yiisoft\\Mailer\\MessageBodyTemplate(\n    $aliases->get($params['yiisoft\/mailer']['messageBodyTemplate']['viewPath']),\n)",
> ```
> 
> The proposal is to skip `Callable` objects in the `buildObjectsCache()` method.
> 
> ```
> "Yiisoft\\Definitions\\CallableDefinition#149": {
> 	"private $callable": "static fn (\\Yiisoft\\Aliases\\Aliases $aliases) => new \\Yiisoft\\Mailer\\MessageBodyTemplate(\n    $aliases->get($params['yiisoft\/mailer']['messageBodyTemplate']['viewPath']),\n)"
> },
> ```

Why?

**Comment by xepozz** [comment]
> After discussion with @olegbaturin, we have suggestion.
> 
> 1. Rename `$depth` parameter in `asJsonObjectsMap()` method to `$arrayDepth`.
> 2. `$arrayDepth` will be mean output depth for array object properties only.
> 3. And therefore, pass `$depth+2` instead of `$depth+1` in `asJsonInternal()`.
> 
> @xepozz WDYT?

I don't mind if the name changed
Why $depth+2 instead of $depth+1?

**Comment by olegbaturin** [comment]
> Why?

It's for your evaluation. If orphan lines for closures `"Closure#147": "..."` are required, just decline this proposal.

**Comment by vjik** [comment]
**Variant 1. Consistent to to `asJson()` method**

Level 0:

```json
"array (1 item) [...]"
```

Level 1:

```json
{
    "stdClass#2512": "stdClass#2512 (...)"
}
```

Level 2:

```json
{
    "stdClass#2512": {
        "public $name": "nested1",
		"public $var": "array (1 item) [...]"
    }
}
```

**Variant 2. Based on object level.**

Level 0:

```json
{
    "stdClass#2512": "stdClass#2512 (...)"
}
```

Level 1:

```json
{
    "stdClass#2512": {
        "public $name": "nested1",
		"public $var": "array (2 items) [...]"
    }
}
```

Level 2:

```json
{
    "stdClass#2512": {
        "public $name": "nested1",
		"public $var": [
			"test",
			"array (2 items) [...]"
		]
    }
}
```

**Variant 3. Based on arrays.**

Level 0:

```json
{
    "stdClass#2512": {
        "public $name": "nested1",
		"public $var": "array (1 item) [...]"
    }
}
```

Level 1:

```json
{
    "stdClass#2512": {
        "public $name": "nested1",
		"public $var": [
			"test",
			"array (2 items) [...]"
		]
    }
}
```

Level 2:

```json
{
    "stdClass#2512": {
        "public $name": "nested1",
		"public $var": [
			"test",
			[
				"hello",
				"world"
			]
		]
    }
}
```

-----

Does a result of `asJsonObjectsMap()` with collapsed objects make sense:

```json
{
    "stdClass#2512": "stdClass#2512 (...)"
}
```

?

If no, we can use "Variant 3. Based on arrays." or "Variant 1. Consistent to to `asJson()` method", but limit minimum value of $depth to two.

**Review by xepozz — APPROVED** [review]

**Pull request merged**